### PR TITLE
feat(agent): add additional arguments as error attributes for newrelic_notice_error

### PR DESCRIPTION
**Pull request by hahuja2**
This PR does the following:

- Create new function called `nr_txn_record_error_with_additional_attributes`, which takes in the additional parameters that are passed in `newrelic_notice_error` and adds them as error attributes.
- Add unit tests to test new functions
- Add integration tests to verify additional parameters are showing up as error attributes

### CODE REVIEW

File: agent/php_api.c [hunk 1]
```diff
@@ -177,6 +181,23 @@ PHP_FUNCTION(newrelic_notice_error) {
...
+    char* stack_json = nr_php_backtrace_to_json(NULL TSRMLS_CC);
+
+    nr_txn_record_error_with_additional_attributes(
+        NRPRG(txn), priority, true, buf1, errclass, buf2,
+        error_line, buf3, error_number, stack_json);
+
+    nr_free(buf1);
+    nr_free(buf2);
+    nr_free(buf3);
+    nr_free(stack_json);
```
Comment (ZNeumann) [inline agent/php_api.c:199]: What is the purpose of copying these strings only to immediately free them? Is there a reason why we cannot just directly pass in `errormsgstr`, `error_file`, and `error_context`?

File: axiom/nr_errors.c [hunk 2]
```diff
@@ -97,6 +170,25 @@ void nr_error_destroy(nr_error_t** error_ptr) {
...
+  if (0 == error) {
+    return;
+  }
+  nr_free(error->message);
+  nr_free(error->klass);
+  nr_free(error->error_file);
+  nr_free(error->error_context);
+  nr_free(error->span_id);
+  nr_free(error->stacktrace_json);
+  nr_realfree((void**)error_ptr);
```
Comment (ZNeumann) [inline axiom/nr_errors.c:189]: Based on the name of this function, I would expect it to only destroy the additional params, and thus still require the calling of `nr_error_destroy` afterward to fully free the error.

File: axiom/nr_errors_private.h [hunk 3]
```diff
@@ -25,6 +25,11 @@ struct _nr_error_t {
   int priority;          /* Error priority - lowest to highest */
   char* message;         /* Error message */
   char* klass;           /* Error class */
+  char* error_file;      /* Error file */
+  int error_line;        /* Error line */
+  char* error_context;   /* Error context */
+  int error_no;          /* Error number */
+  int option;            /* Error option */
```
Comment (ZNeumann) [inline axiom/nr_errors_private.h:32]: As far as I can tell, these are specifically user-provided through the API. It would be useful to specify that in the name of the variables, or at least their comment.

Comment (ZNeumann) [inline axiom/nr_errors_private.h:32]: So am I correct in that the `option` variable indicates whether the other user-provided variables are populated? Could we not just check, for example, `error_file` for null behind an API like `nr_error_has_user_attributes` rather than having this confusing variable? As far as I can tell, these attributes are either all-or-nothing.

File: axiom/nr_txn.c [hunk 4]
```diff
@@ -1638,6 +1641,109 @@ void nr_txn_record_error(nrtxn_t* txn,
   nr_free(span_id);
 }
 
+void nr_txn_record_error_with_additional_attributes(
```
Comment (ZNeumann) [inline axiom/nr_txn.c:1644]: There is a large amount of code duplication in this API. It would be nice to create a more unified API such that we don't have multiple copies of large functions.

File: agent/php_api.c (hunk 1 above)
Comment (hahuja2) [reply agent/php_api.c]: Good point, fixed: [b8e6402](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/b8e64022e5b029a057aece820b663e4b444e556b)

File: axiom/nr_errors.c (hunk 2 above)
Comment (hahuja2) [reply axiom/nr_errors.c]: To reduce code duplication, I have removed this function and added the additional logic in `nr_error_destroy` inside of a if statement: [f917ced](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/f917ced797ecc2b0f2f560487081e73bf4cf47a1)

File: axiom/nr_errors_private.h (hunk 3 above)
Comment (hahuja2) [reply axiom/nr_errors_private.h]: Good point, I have updated the comments: [9f3e7e6](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/9f3e7e652eeeaef0e19c45a1aacbd5ba6891d161)

Comment (hahuja2) [reply axiom/nr_errors_private.h]: Agree, the `option` variable isn't necessary and I have removed it: [9f3e7e6](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/9f3e7e652eeeaef0e19c45a1aacbd5ba6891d161)

File: axiom/nr_txn.c (hunk 4 above)
Comment (hahuja2) [reply axiom/nr_txn.c:1675]: Good point, I have added helper functions for `nr_txn_record_error`, `nr_segment_set_error`, and `nr_error_create` to reduce code duplication: [23823f9](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/23823f9ac52055e35d00951a9af83aadaeb756fe) and [f917ced](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/f917ced797ecc2b0f2f560487081e73bf4cf47a1)

File: agent/php_api.c [hunk 5]
```diff
@@ -144,12 +146,14 @@ PHP_FUNCTION(newrelic_notice_error) {
     case 5:
       if (FAILURE
           == zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET,
-                                      ZEND_NUM_ARGS() TSRMLS_CC, "lsslz!",
-                                      &ignore1, &errormsgstr, &errormsglen,
-                                      &ignore2, &ignore3, &ignore4, &ignore5)) {
+                                      ZEND_NUM_ARGS() TSRMLS_CC, "lssls!",
```
Comment (mfulb) [inline agent/php_api.c:149]: The format for the parameters changes so the last value is a 's' instead of 'z' - what is the rationale behind this change?

File: agent/php_api.c [hunk 6]
```diff
@@ -168,13 +172,23 @@ PHP_FUNCTION(newrelic_notice_error) {
     }
   }
 
-  {
-    char* buf = nr_strndup(errormsgstr, errormsglen);
+  if (!determine) {
```
Comment (mfulb) [inline agent/php_api.c:175]: This code seems to be able to be reduced as there are common elements to each possible path - also use an `else` instead of having two separate `if` statements testing logically opposite conditions.

File: axiom/nr_errors.c [hunk 7]
```diff
@@ -14,37 +14,68 @@
...
+                            const char* error_file,
+                            int error_line,
+                            const char* error_context,
+                            int error_no,
                             const char* stacktrace_json,
                             const char* span_id,
                             nrtime_t when) {
   nr_error_t* error;
 
-  if (0 == message) {
```
Comment (mfulb) [inline axiom/nr_errors.c:25]: Why were these guards removed?

File: axiom/nr_errors.c [hunk 8]
```diff
@@ -14,37 +14,68 @@
...
-    return 0;
-  }
-
   error = (nr_error_t*)nr_zalloc(sizeof(nr_error_t));
   error->priority = priority;
   error->when = when;
   error->message = nr_strdup(message);
   error->klass = nr_strdup(klass);
   error->stacktrace_json = nr_strdup(stacktrace_json);
+  if (NULL != error_file && NULL != error_context) {
```
Comment (mfulb) [inline axiom/nr_errors.c:35]: If only one of these two values is defined do we still want to create the associated field?

File: axiom/nr_errors.c [hunk 9]
```diff
@@ -14,37 +14,68 @@
...
+                                              const char* message,
+                                              const char* klass,
+                                              const char* error_file,
+                                              int error_line,
+                                              const char* error_context,
+                                              int error_no,
+                                              const char* stacktrace_json,
+                                              const char* span_id,
+                                              nrtime_t when) {
+  if (0 == message || 0 == klass || 0 == error_file || 0 == error_context
```
Comment (mfulb) [inline axiom/nr_errors.c:71]: Please change usages in this file of `0` to `NULL` if used to test a pointer value.  The usage of `0` is from old code and we do not use that form now.

File: axiom/nr_errors.c [hunk 10]
```diff
@@ -94,6 +153,10 @@ void nr_error_destroy(nr_error_t** error_ptr) {
   nr_free(error->klass);
   nr_free(error->span_id);
   nr_free(error->stacktrace_json);
+  if ((NULL != error->error_file) && (NULL != error->error_context)) {
```
Comment (mfulb) [inline axiom/nr_errors.c:156]: `nr_free` should handle this case.

File: axiom/nr_errors.h [hunk 11]
```diff
@@ -38,6 +38,38 @@ extern nr_error_t* nr_error_create(int priority,
...
+ *              The PHP agent uses PHP's predefined error constant names.
+ *              Examples are "E_ERROR" and "E_WARNING".
+ *           4. Error file provided by user.
+ *           5. Error line provided by user.
+ *           6. Error context provided by user.
+ *           7. Error number provided by user.
+ *           8. String containing stack trace in JSON format.
+ *           9. Span ID
+ *           10. When the error occurred.
+ *
```
Comment (mfulb) [inline axiom/nr_errors.h:58]: Would be good to explain if all are required or if there are certain combinations allowed.

File: axiom/nr_errors.h [hunk 12]
```diff
@@ -54,6 +86,35 @@ extern const char* nr_error_get_message(const nr_error_t* error);
  * Returns : The klass of the error or NULL on failure.
  */
 extern const char* nr_error_get_klass(const nr_error_t* error);
+
+/*
+ * Purpose : Get the error file of an error.
+ *
+ * Returns : The error file of the error or NULL on failure.
```
Comment (mfulb) [inline axiom/nr_errors.h:93]: Is it a 'failure' or simply it is not defined?

File: axiom/nr_segment.c [hunk 13]
```diff
@@ -1170,23 +1170,59 @@ void nr_segment_record_exception(nr_segment_t* segment,
...
+                                        int error_line,
+                                        char* error_context,
+                                        int error_no) {
   if (NULL == segment->error) {
     segment->error = nr_zalloc(sizeof(nr_segment_error_t));
   }
 
   nr_free(segment->error->error_message);
   nr_free(segment->error->error_class);
+  if (NULL != error_file && NULL != error_context) {
```
Comment (mfulb) [inline axiom/nr_segment.c:1186]: This code seems like it could leak.  If the caller doesn't define these values we don't free the current values.  Below the current value is set to `NULL` if the caller doesn't define a value, thus losing a reference to the memory?

File: axiom/nr_segment.c [hunk 14]
```diff
@@ -1170,23 +1170,59 @@ void nr_segment_record_exception(nr_segment_t* segment,
...
+    nr_free(segment->error->error_file);
+    nr_free(segment->error->error_context);
+  }
 
   segment->error->error_message
       = error_message ? nr_strdup(error_message) : NULL;
   segment->error->error_class = error_class ? nr_strdup(error_class) : NULL;
+  if (NULL != error_file && NULL != error_context) {
```
Comment (mfulb) [inline axiom/nr_segment.c:1194]: I'm not clear on the behavior of this function if the segment had all fields defined previously and then this function is called and `error_file` and `error_context` are `NULL`.  It seems the previous values would be retained.  That could be confusing as it seems there is no way to clear those fields once set.

File: axiom/nr_segment.c [hunk 15]
```diff
@@ -1170,23 +1170,59 @@ void nr_segment_record_exception(nr_segment_t* segment,
...
 }
 
-void nr_segment_set_error(nr_segment_t* segment,
-                          const char* error_message,
-                          const char* error_class) {
-  if ((NULL == segment) || (NULL == error_message && NULL == error_class)) {
-    return;
-  }
-
+static void nr_segment_set_error_helper(nr_segment_t* segment,
```
Comment (mfulb) [inline axiom/nr_segment.c:1173]: I would recommend putting a comment on each of these static helper functions throughout this file to make it clear no NULL checking is done.  It is unlikely any new code would call these directly but I think it is a good safety guard to make sure a coder is aware of this.

Do you think we need unit tests for these helper functions throughout this file?

File: axiom/nr_segment.h [hunk 16]
```diff
@@ -117,6 +117,10 @@ typedef struct _nr_segment_metric_t {
 typedef struct _nr_segment_error_t {
   char* error_message; /* The error message that will appear on a span event. */
   char* error_class;   /* The error class that will appear on a span event. */
+  char* error_file;
```
Comment (mfulb) [inline axiom/nr_segment.h:120]: Can you add a comment for each of these fields like the existing ones have?

File: axiom/nr_txn.c (hunk 4 above)
Comment (mfulb) [reply axiom/nr_txn.c:1675]: Could you remove `nr_txn_record_error_helper` and then just have `nr_txn_record_error` call `nr_txn_record_error_with_additional_attributes` with the unused fields filled with nothing values?

File: axiom/nr_txn.c [hunk 17]
```diff
@@ -2479,6 +2536,13 @@ nr_analytics_event_t* nr_error_to_event(const nrtxn_t* txn) {
   nro_set_hash_string(params, "error.class", nr_error_get_klass(txn->error));
   nro_set_hash_string(params, "error.message",
                       nr_error_get_message(txn->error));
+  if (nr_error_get_file(txn->error) && nr_error_get_context(txn->error)) {
```
Comment (mfulb) [inline axiom/nr_txn.c:2539]: Is it ever possible we'd only one to write a subset of these 4 or do they either all exist or not exist?

File: axiom/nr_txn.h [hunk 18]
```diff
@@ -471,6 +471,17 @@ extern nr_status_t nr_txn_record_error_worthy(const nrtxn_t* txn, int priority);
 #define NR_TXN_ALLOW_RAW_EXCEPTION_MESSAGE \
   "Message removed by New Relic security settings"
 
+extern void nr_txn_record_error_with_additional_attributes(nrtxn_t* txn,
```
Comment (mfulb) [inline axiom/nr_txn.h:474]: Could you document the record error functions (both forms)?

File: axiom/tests/test_txn.c [hunk 19]
```diff
@@ -8098,6 +8098,126 @@ static void test_segment_record_error(void) {
...
+  nr_distributed_trace_set_sampled(txn->distributed_trace, true);
+  txn->options.allow_raw_exception_messages = 1;
+
+  /* No error attributes added if error collection isn't enabled */
+  txn->options.err_enabled = 0;
+  nr_txn_record_error_with_additional_attributes(
+      txn, 1, true, "low priority message", "low priority class", "random.php",
+      150, "random context", 256, "[\"A\",\"B\"]");
+  tlib_pass_if_null("No segment error created", segment->error);
+  txn->options.err_enabled = 1;
```
Comment (mfulb) [inline axiom/tests/test_txn.c:8128]:
```suggestion

  /* Enable error collection */
  txn->options.err_enabled = 1;
```

File: axiom/tests/test_txn.c [hunk 20]
```diff
@@ -8098,6 +8098,126 @@ static void test_segment_record_error(void) {
...
+  segment = nr_segment_start(txn, NULL, NULL);
+  nr_distributed_trace_set_sampled(txn->distributed_trace, true);
+  txn->options.allow_raw_exception_messages = 1;
+
+  /* No error attributes added if error collection isn't enabled */
+  txn->options.err_enabled = 0;
+  nr_txn_record_error_with_additional_attributes(
+      txn, 1, true, "low priority message", "low priority class", "random.php",
+      150, "random context", 256, "[\"A\",\"B\"]");
+  tlib_pass_if_null("No segment error created", segment->error);
```
Comment (mfulb) [inline axiom/tests/test_txn.c:8127]: Shouldn't `txn->error` also be checked that no error was created?

File: tests/integration/attributes/test_transaction_additional_params2.php [hunk 21]
```diff
@@ -0,0 +1,42 @@
+<?php
```
Comment (mfulb) [inline tests/integration/attributes/test_transaction_additional_params2.php:1]: I would add tests passing invalid values to both forms of this API.  For example, passing `NULL` or passing an integer or object where a string is expected, etc.   Make sure a mistake calling from PHP will not take the agent down.

File: agent/php_api.c (hunk 5 above)
Comment (hahuja2) [reply agent/php_api.c:149]: This has been changed to a 's' because in the [docs API](https://docs.newrelic.com/docs/apm/agents/php-agent/php-agent-api/newrelic_notice_error/#syntax), it mentions that `error_context` is a string which is the 5th parameter for this function.

File: axiom/nr_errors.c (hunk 7 above)
Comment (hahuja2) [reply axiom/nr_errors.c:25]: These were removed because these guards are being checked before this helper function is called. An example of this is in [nr_error_create](https://github.com/newrelic/newrelic-php-agent/pull/942/files#diff-6f14a71dbed1d482f4f6006b000174fbd0e41f5138cdb61664ea50e9a1aa7947R54).

File: axiom/tests/test_segment.c [hunk 22]
```diff
@@ -2868,6 +2874,12 @@ static void test_segment_set_error_attributes(void) {
   tlib_pass_if_str_equal("error.class", "error.class 1",
                          segment.error->error_class);
 
+  nr_segment_set_error_with_additional_params(&segment, "error.message", "error.class", "error.file", 125, "rand3", 100);
```
[the remark this answers is not in the document]
Comment (hahuja2) [reply axiom/tests/test_segment.c:2877]: If the user passes in an empty string, then the attribute will still show up. The attribute will not show up if `NULL` is passed in.

File: axiom/nr_errors.c (hunk 8 above)
Comment (hahuja2) [reply axiom/nr_errors.c:35]: After discussion, it has been decided that if one attribute is passed in as `NULL`, the rest of the attributes should still show up. I will create a commit for this.

File: axiom/nr_segment.c (hunk 14 above)
Comment (hahuja2) [reply axiom/nr_segment.c:1194]: After discussion, it has been decided that if `NULL` is passed in, the attribute should not show up. I will modify the code and create a commit for this.

File: agent/php_api.c (hunk 6 above)
Comment (hahuja2) [reply agent/php_api.c]: Fixed: [66a069d](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/66a069dd06cd14f1dc72f87f940a794d8180053a)

File: axiom/nr_errors.c (hunk 9 above)
Comment (hahuja2) [reply axiom/nr_errors.c]: Fixed: [ace780f](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/ace780f45742b8df90d08494c340bda271feb495)

File: axiom/nr_errors.c (hunk 10 above)
Comment (hahuja2) [reply axiom/nr_errors.c]: Fixed: [c058df8](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/c058df86277f602e0f73e37e0a0a86e805a2d7d4)

File: axiom/nr_errors.h (hunk 12 above)
Comment (hahuja2) [reply axiom/nr_errors.h]: Good point, it should instead say that it was not defined, fixed: [9fad7fa](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/9fad7fae54e3fa1a4a2d2dfecf7493e61ee1f2a3)

File: axiom/nr_errors.h (hunk 11 above)
Comment (hahuja2) [reply axiom/nr_errors.h:60]: Added: [9fad7fa](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/9fad7fae54e3fa1a4a2d2dfecf7493e61ee1f2a3)

File: axiom/nr_segment.c (hunk 14 above)
Comment (hahuja2) [reply axiom/nr_segment.c]: I have modified this function where all strings are freed now, so the previous values will not be retained. New values will only be added if they are not `NULL`: [ed191c0](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/ed191c0b4ee2cff587aa3fcee1219c9133e586f7), [bc0a82e](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/bc0a82e066d75632bad2a84be5fa90b5dc42d5ed)

File: axiom/nr_segment.h (hunk 16 above)
Comment (hahuja2) [reply axiom/nr_segment.h]: Added: [eb64129](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/eb6412921bd5a5b301a6f76ffde87aa581ce664e)

File: axiom/nr_txn.h (hunk 18 above)
Comment (hahuja2) [reply axiom/nr_txn.h:480]: Added: [c5a40cc](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/c5a40cc3cabb46abb889e5ba51355091dd73f2e2)

File: axiom/tests/test_segment.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (hahuja2) [reply axiom/tests/test_segment.c]: Added: [06430a2](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/06430a26c136c81d986351c1714e6ae814e9e731)

File: axiom/tests/test_txn.c (hunk 20 above)
Comment (hahuja2) [reply axiom/tests/test_txn.c:8127]: Good point, added: [42a7a81](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/42a7a81482af738ad7a4602ad9babe4cab6becaa)

File: axiom/nr_txn.c (hunk 4 above)
Comment (hahuja2) [reply axiom/nr_txn.c:1555]: Good idea, I have removed all of the helper functions: [e177f40](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/e177f40a8b868410e5e1b728df7b5df8356ce1f4), [ed191c0](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/ed191c0b4ee2cff587aa3fcee1219c9133e586f7), [788c5a0](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/788c5a0e6330c7c788b2a6c5b028391319b7af17)

File: axiom/tests/test_txn.c [hunk 23]
```diff
@@ -8098,6 +8098,126 @@ static void test_segment_record_error(void) {
...
+  tlib_pass_if_int_equal("Correct segment error.no", 256,
+                         segment->error->error_no);
+  tlib_pass_if_str_equal("txn error message matches segment error message",
+                         segment->error->error_message,
+                         nr_error_get_message(txn->error));
+  tlib_pass_if_str_equal("txn error class matches segment error class",
+                         segment->error->error_class,
+                         nr_error_get_klass(txn->error));
+
+  /* Multiple errors on the same segment */
```
[the remark this answers is not in the document]
Comment (hahuja2) [reply axiom/tests/test_txn.c]: Added: [42a7a81](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/42a7a81482af738ad7a4602ad9babe4cab6becaa)

File: axiom/nr_segment.c (hunk 15 above)
Comment (hahuja2) [reply axiom/nr_segment.c]: I have removed all of the helper functions: [e177f40](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/e177f40a8b868410e5e1b728df7b5df8356ce1f4), [ed191c0](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/ed191c0b4ee2cff587aa3fcee1219c9133e586f7), [788c5a0](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/788c5a0e6330c7c788b2a6c5b028391319b7af17)

**Comment by mfulb** [comment]
PR #960 supersedes the work in this PR.  The conclusion was that the agent spec does not support these current parameters currently as attributes, and much of the information is available from the stack trace or by adding the desired information to the error message passed to the API.
